File: stian/day_three/python/day_three.py
import aoc_frame

def part_one(data, flag) :
    count_zero = [0 for i in range(len(data[0]))]
    count_one = [0 for i in range(len(data[0]))]
    for i in data :
        for k, v in enumerate(i) :
            if v == '1' :
                count_one[k] += 1
            else :
                count_zero[k] += 1
    
    gamma = ''
    epsilon = ''
    for i in range(len(count_zero)) :
        if count_zero[i] > count_one[i] :
            gamma += '0'
            epsilon += '1'
        elif count_zero[i] < count_one[i] :
            gamma += '1'
            epsilon += '0'
        else :
            gamma += '1'
            epsilon += '0'
    
    if flag :
        return gamma, epsilon

    gamma = int(gamma, 2)
    epsilon = int(epsilon, 2)

    return gamma * epsilon

def part_two(data) :                
    gamma, epsilon = part_one(data, True)
    kept = data
    i = 0
    while len(kept) > 1 :
        gamma, epsilon = part_one(kept, True)
        kept = find_remaining(kept, gamma, i)
        i += 1
    oxygen = int(kept[0], 2)
    
    kept = data
    i = 0
    while len(kept) > 1 :
        gamma, epsilon = part_one(kept, True)
        kept = find_remaining(kept, epsilon, i)
        i += 1
    
    carbon = int(kept[0], 2)

    return oxygen * carbon

def find_remaining(data, keyword, index) :
    new_data = []
    for i in data :
        if i[index] == keyword[index] :
            new_data.append(i)
    
    return new_data

if __name__ == '__main__' :
    data = []
    test_data = []

    data, test_data = aoc_frame.read_data(2021, 3)
    print('Test Data:', test_data, '\n')
    aoc_frame.format_strings(part_one(data, False), part_one(test_data, False), 1)
    aoc_frame.format_strings(part_two(data), part_two(test_data), 2)

File: stian/day_three/go/day_three.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	file, err := os.Open("day_three.txt")
	if err != nil {
		panic(err)
	}

	defer file.Close()
	scanner := bufio.NewScanner(file)
	var data []string

	for scanner.Scan() {
		data = append(data, scanner.Text())
	}

	fmt.Println("Part 1:", partOne(data))
	fmt.Println("Part 2:", partTwo(data))
}

func partOne(data []string) int64 {
	count_ones := make([]int, len(data[0]))
	count_zero := make([]int, len(data[0]))
	for _, val := range data {
		for i, v := range val {
			if string(v) == "0" {
				count_zero[i]++
			} else if string(v) == "1" {
				count_ones[i]++
			}
		}
	}
	num1 := ""
	num2 := ""
	for i := 0; i < len(count_ones); i++ {
		if count_ones[i] > count_zero[i] {
			num1 += "1"
			num2 += "0"
		} else {
			num1 += "0"
			num2 += "1"
		}
	}
	gamma, _ := strconv.ParseInt(num1, 2, 64)
	epsilon, _ := strconv.ParseInt(num2, 2, 64)

	return gamma * epsilon
}

func findCommon(data []string, keyword string, index int) []string {
	tmp := make([]string, 0)
	for _, val := range data {
		if val[index] == keyword[index] {
			tmp = append(tmp, val)
		}
	}
	return tmp
}

func findGammaEpisolon(data []string) (string, string) {
	count_ones := make([]int, len(data[0]))
	count_zero := make([]int, len(data[0]))
	for _, val := range data {
		for i, v := range val {
			if string(v) == "0" {
				count_zero[i]++
			} else if string(v) == "1" {
				count_ones[i]++
			}
		}
	}
	gamma := ""
	epsilon := ""
	for i := 0; i < len(count_ones); i++ {
		if count_ones[i] > count_zero[i] {
			gamma += "1"
			epsilon += "0"
		} else if count_ones[i] < count_zero[i] {
			gamma += "0"
			epsilon += "1"
		} else {
			// the counts are equal
			gamma += "1"
			epsilon += "0"
		}
	}
	return gamma, epsilon
}

func partTwo(data []string) int64 {
	nums := data
	i := 0
	for len(nums) > 1 {
		gamma, _ := findGammaEpisolon(nums)
		nums = findCommon(nums, gamma, i)
		i++
	}
	oxygen, _ := strconv.ParseInt(nums[0], 2, 64)
	nums = data
	i = 0
	for len(nums) > 1 {
		_, epsilon := findGammaEpisolon(nums)
		nums = findCommon(nums, epsilon, i)
		i++
	}
	carbon, _ := strconv.ParseInt(nums[0], 2, 64)
	return oxygen * carbon
}
